src/passes/reconstruct-stringify-walker.cpp

```diff
@@ -0,0 +1,193 @@
+#include "stringify-walker.h"
+
+#if RECONSTRUCT_DEBUG
+#define DBG(statement) statement
+#else
+#define DBG(statement)
+#endif
+
+namespace wasm {
+
+ReconstructStringifyWalker::ReconstructStringifyWalker(Module* wasm)
+  : existingBuilder(*wasm), outlinedBuilder(*wasm) {
+  this->setModule(wasm);
+  DBG(std::cout << "\n\nexistingBuilder: " << &existingBuilder
+                << " outlinedBuilder: " << &outlinedBuilder);
+}
+
+void ReconstructStringifyWalker::addUniqueSymbol(SeparatorReason reason) {
+  if (auto curr = reason.getFuncStart()) {
+    startExistingFunction(curr->func);
+    return;
+  }
+
+  // instrCounter is managed manually and incremented at the beginning of
+  // addUniqueSymbol() and visitExpression(), except for the case where we are
+  // starting a new function, as that resets the counters back to 0
+  instrCounter++;
+  if (state == InSkipSeq) {
+    return;
+  }
+
+  DBG(std::string desc);
+  IRBuilder* builder = state == InSeq ? &outlinedBuilder : &existingBuilder;
+  if (auto curr = reason.getBlockStart()) {
+    if (auto expr = builder->visitBlockStart(curr->block);
+        auto err = expr.getErr()) {
+      Fatal() << err->msg;
+    }
+    DBG(desc = "Block Start for ";);
+  } else if (auto curr = reason.getIfStart()) {
+    if (auto expr = builder->visitIfStart(curr->iff);
+        auto err = expr.getErr()) {
+      Fatal() << err->msg;
+    }
+    DBG(desc = "If Start for ";);
+  } else if (reason.getEnd()) {
+    if (auto expr = builder->visitEnd(); auto err = expr.getErr()) {
+      Fatal() << err->msg;
+    }
+    DBG(desc = "End for ";);
+  } else {
+    DBG(desc = "addUniqueSymbol for unhandled instruction ";);
```

Comment:
We could add a `WASM_UNREACHABLE` here, too.

Reply:
done